### PR TITLE
email-templates: add `renderAll`, make `locals` optional

### PR DESCRIPTION
**Pull request by qqilihq**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
* https://github.com/niftylettuce/email-templates#render-html-andor-text
* https://github.com/niftylettuce/email-templates/blob/af01d14fef9df70370f82761d7116ccb979edc2c/src/index.js#L186
* https://github.com/niftylettuce/email-templates/blob/af01d14fef9df70370f82761d7116ccb979edc2c/src/index.js#L135
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/email-templates/index.d.ts [hunk 1]
```diff
@@ -103,11 +112,26 @@ declare namespace EmailTemplate {
...
+         * @param view Name of the template
+         * @param locals The template variables
+         */
+        function renderAll(view: string, locals?: any): Promise<EmailMessage>;
 
         /**
          * Send the Email
          */
         function send(options: EmailOptions): any;
+
```
Comment (blankstar85) [inline types/email-templates/index.d.ts]: Interface EmailMessage should be declared with the rest of the interfaces.

File: types/email-templates/index.d.ts [hunk 2]
```diff
@@ -103,11 +112,26 @@ declare namespace EmailTemplate {
...
+         *
+         * @param view Name of the template
+         * @param locals The template variables
+         */
+        function renderAll(view: string, locals?: any): Promise<EmailMessage>;
 
         /**
          * Send the Email
          */
         function send(options: EmailOptions): any;
```
Comment (blankstar85) [inline types/email-templates/index.d.ts]: This should return the EmailMessage as well

**Review by blankstar85 — CHANGES_REQUESTED** [review]
Looks good outside of the two changes.

File: types/email-templates/index.d.ts (hunk 2 above)
Comment (qqilihq) [reply types/email-templates/index.d.ts]: Looking at this part, it's rather a `Promise` of `SentMessageInfo` from nodemailer (which is currently just defined as `any` in the nodemailer typings) and **not** the newly defined interface `EmailMessage`.

See:  https://github.com/niftylettuce/email-templates/blob/af01d14fef9df70370f82761d7116ccb979edc2c/src/index.js#L292

I'll thus change it to return `Promise<any>` instead.

File: types/email-templates/index.d.ts (hunk 1 above)
Comment (qqilihq) [reply types/email-templates/index.d.ts]: Not sure how to do this … if I simply move it to the upper level, it seems I can no longer access it from outside?

**Comment by qqilihq** [comment]
@gurisko @blankstar85 -- any feedback?

**Comment by qqilihq** [comment]
@cyrilschumacher @gurisko @blankstar85  -- any feedback?

File: types/email-templates/index.d.ts [hunk 3]
```diff
@@ -84,7 +85,15 @@ declare class EmailTemplate {
...
-    render(view: string, locals: any): Promise<string>;
+    render(view: string, locals?: any): Promise<string>;
+    /**
+     * Render all available template files for a given email
+     * template (e.g. `html.pug`, `text.pug`, and `subject.pug`)
+     *
+     * @param view Name of the template
+     * @param locals The template variables
+     */
+    renderAll(view: string, locals?: any): Promise<EmailTemplate.EmailMessage>;
```
Comment (blankstar85) [inline types/email-templates/index.d.ts]: remove EmailTemplate and just use EmailMessage, after the interface is moved with the others.

File: types/email-templates/index.d.ts (hunk 1 above)
Comment (blankstar85) [reply types/email-templates/index.d.ts]: You should be able to access from the outside, just like the other interfaces.

**Review by blankstar85 — COMMENTED** [review]
review comments

**Review by blankstar85 — CHANGES_REQUESTED** [review]
Also, bump the version, this still keeps it inline with the latest version of email-templates.

**Pull request closed without merging**